packages/client-utils/src/vstorage-client.js

```diff
@@ -0,0 +1,423 @@
+ * @typedef {{
+ *  latest(height?: Height): Promise<Update<T>>;
+ * }} Topic<T>
+ */
+
+/**
+ * @template {any} T
+ * @typedef {Topic<T> & {
+ *  iterate(minimum?: Height, maximum?: Height): AsyncIterable<Update<T>>;
+ *  reverseIterate(maximum?: Height, minimum?: Height): AsyncIterable<Update<T>>;
+ * }} StreamTopic<T>
+ */
+
+/**
+ * @template {any} T
+ * @typedef {{
+ *  blockHeight: Height;
+ *  value: T;
+ * }} Update<T>
+ */
+
+export const INVALID_HEIGHT_ERROR_MESSAGE = 'Invalid height';
+export const INVALID_RPC_ADDRESS_ERROR_MESSAGE =
+  'No valid RPC address found in config';
+
+const MINIMUM_HEIGHT = 1n;
+
+export const PATH_PREFIX = '/agoric.vstorage.Query';
+
+export const PATHS = /** @type {const} */ ({
+  CHILDREN: 'Children',
+  DATA: 'Data',
+});
+
+const codecs = /** @type {const} */ ({
+  [PATHS.CHILDREN]: {
+    request: QueryChildrenRequest,
+    response: QueryChildrenResponse,
+  },
+  [PATHS.DATA]: {
+    request: QueryDataRequest,
+    response: QueryDataResponse,
+  },
+});
+
+/**
+ * @param {ReturnType<typeof parseValue>} value
+ */
+const assertValueStructure = value => {
+  if (!isStreamCell(value))
+    throw Error(`Expected a stream cell, got ${quote(value)}`);
+};
+
+/**
+ * @param {ReturnType<parseValue>} cell
+ */
+const isDataString = cell => !!cell && typeof cell === 'string';
+
+/**
+ * @param {{
+ *  errorMessage: string;
+ *  height?: Height;
+ *  kind: PATHS[keyof PATHS];
+ *  path: string;
+ *  rpcAddress?: MinimalNetworkConfig['rpcAddrs'][0];
+ * }} options
+ */
+const makeVstorageError = ({ errorMessage, height, kind, path, rpcAddress }) =>
+  makeError(
+    `Cannot read '${kind}' of '${path}' ${
+      rpcAddress ? `from '${rpcAddress}'` : ''
+    } ${height ? `at height '${height}'` : ''} due to error: ${errorMessage}`,
+  );
+
+/**
+ * @template {any} T
+ * @param {{queryClient: ReturnType<typeof makeQueryClient>}} powers
+ * @param {string} path
+ * @param {Partial<{ compat: boolean }>} [options]
+ * @returns {Topic<T>}
+ */
+const makeBlockTopic = ({ queryClient }, path, options) => ({
+  latest: async height => {
+    if (height && height < MINIMUM_HEIGHT)
+      throw makeVstorageError({
+        errorMessage: `${INVALID_HEIGHT_ERROR_MESSAGE} ${height}`,
+        height,
+        kind: PATHS.DATA,
+        path,
+      });
+
+    const response = await queryClient.queryAbci(path, {
+      height: height && height > MINIMUM_HEIGHT ? height - 1n : height,
+      kind: PATHS.DATA,
+    });
+    const parsedValue = parseValue(response.value);
+
+    if (isDataString(parsedValue))
+      return /** @type {Update<T>} */ ({
+        blockHeight: height,
+        value: parsedValue,
+      });
+
+    if (options?.compat) {
+      assertValueStructure(parsedValue);
+      const data = /** @type {StreamCell} */ (parsedValue);
+      const blockHeight = BigInt(data.blockHeight);
+      return /** @type {Update<T>} */ ({
+        blockHeight,
+        value: data.values.reverse().find(Boolean),
```

Comment:
What is the purpose of `.find(Boolean)` against a known StreamCell? Every item in `values` should be relevant, even if it is empty.
```suggestion
        value: data.values.at(-1),
```
(it's true that emptiness might come into play for representing deletion, but even then it would be a mistake to just ignore the deletion)